chain/sync_manager.go

```diff
@@ -52,108 +47,323 @@ type SyncManager interface {
+	})
+
+	for _, ts := range peerHeads {
+		buckets.Insert(ts)
+	}
+
+	if len(buckets.buckets) > 1 {
+		log.Warn("caution, multiple distinct chains seen during head selections")
+		// TODO: we *could* refuse to sync here without user intervention.
+		// For now, just select the best cluster
+	}
+
+	return buckets.Heaviest(), nil
+}
+
+// adds a tipset to the potential sync targets; returns true if there is a a tipset to work on.
+// this could be either a restart, eg because there is no currently scheduled sync work or a worker
+// failed or a potential fork.
+func (sm *syncManager) addSyncTarget(ts *types.TipSet) (*types.TipSet, bool, error) {
+	// Note: we don't need the state lock here to access the active worker states, as the only
+	//       competing threads that may access it do so through State() which is read only.
+
+	// if the worker set is empty, we have finished syncing and were waiting for the next tipset
+	// in this case, we just return the tipset as work to be done
+	if len(sm.state) == 0 {
+		return ts, true, nil
+	}
+
+	// check if it is related to any active sync; if so insert into the pending sync queue
+	for _, ws := range sm.state {
+		if ts.Equals(ws.ts) {
+			// ignore it, we are already syncing it
+			return nil, false, nil
+		}
+
+		if ts.Parents() == ws.ts.Key() {
+			// schedule for syncing next; it's an extension of an active sync
+			sm.pend.Insert(ts)
+			return nil, false, nil
+		}
+	}
+
+	// check to see if it is related to any pending sync; if so insert it into the pending sync queue
+	if sm.pend.RelatedToAny(ts) {
+		sm.pend.Insert(ts)
+		return nil, false, nil
+	}
+
+	// it's not related to any active or pending sync; this could be a fork in which case we
+	// start a new worker to sync it, if it is *heavier* than any active or pending set;
+	// if it is not, we ignore it.
+	for _, ws := range sm.state {
+		if isHeavier(ws.ts, ts) {
+			return nil, false, nil
+		}
+	}
+
+	pendHeaviest := sm.pend.Heaviest()
+	if pendHeaviest != nil && isHeavier(pendHeaviest, ts) {
+		return nil, false, nil
+	}
+
+	// start a new worker, seems heavy enough and unrelated to active or pending syncs
+	return ts, true, nil
+}
+
+// selects the next sync target after a worker sync has finished; returns true and a target
+// TipSet if this chain should continue to sync because there is a heavier related tipset.
+func (sm *syncManager) selectSyncTarget(done *types.TipSet) (*types.TipSet, bool, error) {
+	// we pop the related bucket and if there is any related tipset, we work on the heaviest one next
+	// if we are not already working on a heavier tipset
+	related := sm.pend.PopRelated(done)
+	if related == nil {
+		return nil, false, nil
 	}
-	return ret
+
+	heaviest := related.heaviestTipSet()
+	for _, ws := range sm.state {
+		if isHeavier(ws.ts, heaviest) {
+			return nil, false, nil
+		}
+	}
+
+	return heaviest, true, nil
 }
 
+func isHeavier(a, b *types.TipSet) bool {
+	return a.ParentWeight().GreaterThan(b.ParentWeight())
```

Comment:
we should probably do some other heuristic that accounts for number of blocks in tispets with the same ParentWeight, but idk if that will really affect anything

Reply:
let's think about this a bit, it's not clear if it is the right thing to do.